WordPress/gutenberg
# Try/saving to theme files

## What?
Closes #39194

## Why?
To allow the site editor to behave like a traditional visual builder. A theme can be exported, but when developing from scratch a better devx is to update your theme's files with the changes made via the site editor.

## How?
The following updates are needed:

- added a theme update button (one commit adds it to the more menu, one to the header)
- added a new endpoint to save to theme files the customized templates and template parts
- added a new block editor setting `themeDevMode` based on a new `wp-config.php` constant `THEME_DEV_MODE`

## Testing Instructions
1. Switch to this PR
2. Add `define( 'ENABLE_THEME_SAVING', true );` to your `wp-config.php`
3. Make sure you are on a block theme
4. Make sure WordPress can write to your theme files
5. Go edit a template or template part
6. Save the edit
7. You should see an update theme button. Click it.
8. The updates should be reflected in the files of your current theme

## Screenshots or screencast 

### More menu


https://user-images.githubusercontent.com/107534/163599216-f61d1695-0d4a-45c6-9284-919070c7705a.mp4


### Editor header

https://user-images.githubusercontent.com/107534/163599206-9668fa42-1616-4247-ae25-06f961a4b624.mp4


## Work in progress, known problems

- [ ] check for file writing permissions when adding the editor "theme dev" mode setting
- [ ] what happens if the theme files are updated after the customisations are made? Should we make a backup, just in case?
- [x] [bug?]  I have not fixed the fact that once the customisations are cleared following a theme save, the editor wants to save the clearing of customisations itself.
- [x] [bug] when switching editors, the update theme button only appears once an edit has been done and saved
- [ ] Think about what happens when someone tries to update a directory theme that has had saved customization.

Based heavily on the [create-block-theme](https://github.com/Automattic/create-block-theme/blob/0b30ee424b8b476f3634a43f095340372bcc6922/admin/class-create-block-theme-admin.php) plugin's implementation of this similar concept by @pbking
